tests/flags/flags.go

```diff
@@ -41,6 +41,7 @@ var PreviousReleaseTag = ""
 var PreviousReleaseRegistry = ""
 var ConfigFile = ""
 var SkipShasumCheck bool
+var IsDualStackedNetworkConfig bool
```

Comment:
Can we detect if we are dual stack capable with some introspection ? like checking if we have two ips or something like that, so we don't have to pass flags.

Reply:
See [here](https://github.com/kubevirt/kubevirt/pull/3989#discussion_r475475243) for an explanation on why we're taking the 'flag' route.

Reply:
ack

Reply:
I read the explanation. I agree we shouldn't use an env var checking what provider we are using.
But why not checking whether a known pod (virt-handler for example) has 2 ips (ipv4 and ipv6)?

Reply:
Because in this case a bug on the cluster that cause it to have just one ip, will be masked forever i think,
its good to have an external hint, and this hint should work also on external providers as Edy wrote in the ref link.

Reply:
Checking the ips will work for external providers as well.
We can add to the kubevirtci a check that a dual stack cluster provides two ips or any other required tests.
The kubevirt e2e tests should assume the cluster they are using is working correctly.

Reply:
this is what i suggested for kubevirtci.
but it wont let us know an external provider is working properly.

Reply:
We can document that external has to be dual stack or tests will fail.

Reply:
The owner of the cluster exetrnal or internal should take care for its validity, not the tests using it.

Reply:
> Checking the ips will work for external providers as well.
> We can add to the kubevirtci a check that a dual stack cluster provides two ips or any other required tests.

kubevirtci is not the only cluster against which these tests are executed. We have external clusters of openshift and partners with their own stuff.

> The kubevirt e2e tests should assume the cluster they are using is working correctly.

e2e tests assume little because their exist to check integration with k8s itself, with other addons which we do not control in kubevirt and part of the e2e are to be used as certifications/acceptance for partners that want to check if their own stuff works with kubevirt.

Reply:
imo its wrong, we might miss bugs on downstream or so on, in the future.

Reply:
> > Checking the ips will work for external providers as well.
> > We can add to the kubevirtci a check that a dual stack cluster provides two ips or any other required tests.
> 
> kubevirtci is not the only cluster against which these tests are executed. We have external clusters of openshift and partners with their own stuff.

So what? Wouldn't checking the ips work for those clusters?

> 
> > The kubevirt e2e tests should assume the cluster they are using is working correctly.
> 
> e2e tests assume little because they exist to check integration with k8s itself, with other addons which we do not control in kubevirt and part of the e2e are to be used as certifications/acceptance for partners that want to check if their own stuff works with kubevirt.

We can add acceptance test for dual stack cluster (although I don't think we were requested for such a thing). But having a flag that all our networking tests are using. Just for the case that somebody want to check his dual stack cluster, but the cluster actually doesn't even have two ips on the pods, sounds to me overkill and wrong.

Reply:
> imo its wrong, we might miss bugs on downstream or so on, in the future.

Can you please explain how we will miss bugs in the future?

Reply:
> > imo its wrong, we might miss bugs on downstream or so on, in the future.
> 
> Can you please explain how we will miss bugs in the future?

As edy said 
https://github.com/kubevirt/kubevirt/pull/4035#discussion_r481991140

If we check on external provider how many ips there, and according that not check dual stack features, 
we wont cover the whole matrix that check how all the components work together.
Its like having an oracle that tell you what you should check, and you count its objective.

Reply:
What if the external cluster creator actually remembered to create a dual stack cluster but forgot (didn't know about) to use our dual stack flag? (That's sounds to me like a more realistic scenario).

Reply:
Then he cant blame us xD 
so lets add double check if so ?
both this flag and the check actual pod status

Reply:
@AlonaKaplan , this is a matter of controlling the intent of what the tests need to check.
The tests are written at the moment in a generic way such that it will cover both single or dual stack deployment. However, it does not know to differentiate if one exists or does not, it just checks what is there.

The control part comes from the fact that the user which runs the test knows in advance (based on what was deployed by him) that the expected setup should or should not have dual stack.
From my point of view, it is enough to have a single configuration/state test that based on this user input confirms it is there or not (e.g. given the flag of dual-stack is set, confirm that there are two IP addresses to the Pod).

If this user/test-runner input is missing, we have no way of knowing if the tests pass with or without dual stack.
As you suggested to have a test at the kubevirtci lane, I suggest to have it at the kubevirt tests so we can cover all tested targets, including external clusters.
There is no real need to check against the flag all the time, it may just be a convenience like done in this PR.

I think that the core part of this argument is that you cannot really trust the test target as it is the part we are testing. The system under test is the thing we check, therefore do not trust. The tests and their configuration (e.g. use input) is the part we do trust as it includes our expectation of the behavior.

> What if the external cluster creator actually remembered to create a dual stack cluster but forgot (didn't know about) to use our dual stack flag? (That's sounds to me like a more realistic scenario).

For the tests which run against our kubevirtci providers, logic is in place to enable or disable the flag.
When running on a different cluster, then it is up to the user indeed. I think that the golden test which targets the configuration/status and is dependent on the flag, can as @oshoval mentioned above, check and confirm the lack of it (e.g. that there is only a single IP in place and not two per Pod).

Reply:
Having this flag we will end up with external cluster testers that will have dual stack cluster and won't know the ipv6 flow is not tested. I'm pretty sure nobody will remember to use this flag.
We shouldn't test the dual stack cluster has two ips. Our tests should assume the cluster is fine. We are testing kubevirt flows not kubernetes.
"The system under test is the thing we check" - we check that kubevirt code is working as expected on the system under test. Having two ips per pod on dual stack cluster is not related to kubevirt.